File: tasks/2023/online3/colors.cpp
// @check-accepted: examples brute Nsmall no-limits
#include <iostream>
#include <map>

using namespace std;

long long mod = 1'000'000'007;

map<long long, int> cnt;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int n, k;
    cin >> n >> k;

    cnt[0] = k;

    long long ans = 1;

    for (int i = 0; i < n; i++) {
        long long x;
        cin >> x;
        ans = (ans * cnt[x]) % mod;
        cnt[x + 1]++;
        cnt[x]--;
    }

    cout << ans << endl;
    return 0;
}

File: tasks/2023/online3/colors.py
#!/usr/bin/env python3
# @check-accepted: examples brute Nsmall no-limits
# insert brief description of the solution here


# input data
mod = 10**9 + 7
N, K = map(int, input().strip().split())
V = list(map(int, input().strip().split()))

cnt = [0] * N
cnt[0] = K
ans = 1

for x in V:
    ans = (ans * cnt[x]) % mod
    cnt[x] -= 1
    cnt[x + 1] += 1

print(ans)  # print the result
